mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,547 @@
+        wlsModel.coefficients, wlsModel.intercept).setParent(this))
+      return model
+    }
+
+    val newInstances = instances.map { instance =>
+      val mu = familyLink.initialize(instance.label, instance.weight)
+      val eta = familyLink.predict(mu)
+      Instance(eta, instance.weight, instance.features)
+    }
+
+    val initialModel = new WeightedLeastSquares($(fitIntercept), $(regParam),
+      standardizeFeatures = true, standardizeLabel = true).fit(newInstances)
+
+    val reweightFunc: (Instance, WeightedLeastSquaresModel) => (Double, Double) = {
+      (instance: Instance, model: WeightedLeastSquaresModel) => {
+        val eta = model.predict(instance.features)
+        val mu = familyLink.fitted(eta)
+        val z = familyLink.adjusted(instance.label, mu, eta)
+        val w = familyLink.weights(mu) * instance.weight
+        (z, w)
+      }
+    }
+
+    val optimizer = new IterativelyReweightedLeastSquares(initialModel, reweightFunc,
+      $(fitIntercept), $(regParam), $(maxIter), $(tol))
+
+    val irlsModel = optimizer.fit(instances)
+
+    val model = copyValues(new GeneralizedLinearRegressionModel(uid,
+      irlsModel.coefficients, irlsModel.intercept).setParent(this))
+    model
+  }
+
+  @Since("2.0.0")
+  override def copy(extra: ParamMap): GeneralizedLinearRegression = defaultCopy(extra)
+}
+
+@Since("2.0.0")
+object GeneralizedLinearRegression {
+
+  /** Set of families that GeneralizedLinearRegression supports */
+  private[ml] val supportedFamilies = Set("gaussian", "binomial", "poisson", "gamma")
+
+  /** Set of links that GeneralizedLinearRegression supports */
+  private[ml] val supportedLinks = Set("identity", "log", "inverse", "logit", "probit",
+    "cloglog", "sqrt")
+
+  /** Set of family and link pairs that GeneralizedLinearRegression supports */
+  private[ml] val supportedFamilyLinkPairs = Set(
+    "gaussian" -> "identity", "gaussian" -> "log", "gaussian" -> "inverse",
+    "binomial" -> "logit", "binomial" -> "probit", "binomial" -> "cloglog",
+    "poisson" -> "log", "poisson" -> "identity", "poisson" -> "sqrt",
+    "gamma" -> "inverse", "gamma" -> "identity", "gamma" -> "log"
+  )
+}
+
+/**
+ * :: Experimental ::
+ * Model produced by [[GeneralizedLinearRegression]].
+ */
+@Experimental
+@Since("2.0.0")
+class GeneralizedLinearRegressionModel private[ml] (
+    @Since("2.0.0") override val uid: String,
+    @Since("2.0.0") val coefficients: Vector,
+    @Since("2.0.0") val intercept: Double)
+  extends RegressionModel[Vector, GeneralizedLinearRegressionModel]
+  with GeneralizedLinearRegressionParams {
+
+  private lazy val familyLink = $(family) match {
+    case "gaussian" => if (isDefined(link)) Gaussian($(link)) else Gaussian("identity")
+    case "binomial" => if (isDefined(link)) Binomial($(link)) else Binomial("logit")
+    case "poisson" => if (isDefined(link)) Poisson($(link)) else Poisson("log")
+    case "gamma" => if (isDefined(link)) Gamma($(link)) else Gamma("inverse")
+  }
```

Comment:
This is duplicated code. We can factor this out to `GLMBase`, `Family`, and `Link`.